zkvm/src/txlog.rs

```diff
@@ -118,3 +167,142 @@ impl Entry {
         }
     }
 }
+
+impl MerkleTree {
+    /// Constructs a new MerkleTree based on the input list of entries.
+    pub fn new(list: &[Entry]) -> Option<MerkleTree> {
+        if list.len() == 0 {
+            return None;
+        }
+        let t = Transcript::new(b"ZkVM.txid");
+        Some(MerkleTree {
+            size: list.len(),
+            root: MerkleNode::build_tree(t, list),
+        })
+    }
+
+    /// Builds a proof of inclusion for entry at the given index for the Merkle tree.
+    pub fn proof(&self, index: usize) -> Result<Vec<MerkleNeighbor>, VMError> {
+        if index >= self.size {
+            return Err(VMError::InvalidMerkleProof);
+        }
+        let t = Transcript::new(b"ZkVM.txid");
+        let mut result = Vec::new();
+        self.root.subproof(t, index, self.size, &mut result);
+        Ok(result)
+    }
+}
+
+impl MerkleNode {
+    fn build_tree(mut t: Transcript, list: &[Entry]) -> Self {
+        match list.len() {
+            0 => {
+                let mut leaf = [0u8; 32];
+                TxID::empty(t, &mut leaf);
+                return MerkleNode::Leaf(leaf);
+            }
+            1 => {
+                let mut leaf = [0u8; 32];
+                TxID::leaf(t, &list[0], &mut leaf);
+                return MerkleNode::Leaf(leaf);
+            }
+            n => {
+                let k = n.next_power_of_two() / 2;
+                let mut node = [0u8; 32];
+                let left = Self::build_tree(t.clone(), &list[..k]);
+                let right = Self::build_tree(t.clone(), &list[k..]);
+                t.commit_bytes(b"L", left.hash());
+                t.commit_bytes(b"R", right.hash());
+                t.challenge_bytes(b"merkle.node", &mut node);
+                return MerkleNode::Node(node, Box::new(left), Box::new(right));
+            }
+        }
+    }
+
+    fn subproof(&self, t: Transcript, index: usize, size: usize, result: &mut Vec<MerkleNeighbor>) {
+        let k = size.next_power_of_two() / 2;
+        if index >= k {
```

Comment:
maybe it'll be clearer if you transposed this if/else and matches? Meaning, match first, and if Leaf - return. Then, inside the ::Node case check which neighbour you need and drill in accordingly.

Reply:
ah, yes! doing it that way is much nicer :)